qualtran/bloqs/state_preparation/state_preparation_alias_sampling.py

```diff
@@ -235,6 +235,219 @@ def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+    Signature:
+        selection: The input/output register $|\mathrm{ind}_l\rangle$ of size lg(L) where the desired
+            coefficient state is prepared.
+        temp: Work space comprised of sub signature:
+            - sigma: A mu-sized register containing uniform probabilities for comparison against
+                `keep`.
+            - alt: A lg(L)-sized register of alternate indices
+            - keep: a mu-sized register of probabilities of keeping the initially sampled index.
+            - one bit for the result of the comparison.
+
+    This gate corresponds to the following operations:
+     - UNIFORM_L on the sparse_index register
+     - H^mu on the sigma register
+     - QROM addressed by the selection register into the alt and keep signature.
+     - LessThanEqualGate comparing the keep and sigma signature.
+     - Coherent swap between the selection register and alt register if the comparison
+       returns True.
+
+    Total space will be (2 * log(L) + 2 mu + 1) work qubits + log(L) ancillas for QROM.
+    The 1 ancilla in work qubits is for the `LessThanEqualGate` followed by coherent swap.
+
+    References:
+        [1] [Qubitization of Arbitrary Basis Quantum Chemistry Leveraging Sparsity and Low Rank Factorization](https://arxiv.org/pdf/1902.02134#page=15.30)
+        Berry et al. (2019). Section 5, Eqs. 43, 44.
+        [2] [Encoding Electronic Spectra in Quantum Circuits with Linear T Complexity](https://arxiv.org/abs/1805.03662).
+        Babbush et al. (2018). Section III.D. and Figure 11.
+    """
+    selection_registers: Tuple[Register, ...] = attrs.field(
+        converter=lambda v: (v,) if isinstance(v, Register) else tuple(v)
+    )
+    index: Union[Shaped, NDArray[np.int_]] = attrs.field(eq=_data_or_shape_to_tuple)
+    alt: Union[Shaped, NDArray[np.int_]] = attrs.field(eq=_data_or_shape_to_tuple)
+    keep: Union[Shaped, NDArray[np.int_]] = attrs.field(eq=_data_or_shape_to_tuple)
+    mu: SymbolicInt
+    sum_of_lcu_coeffs: SymbolicFloat
+
+    @classmethod
+    def from_lcu_probs(
+        cls,
+        lcu_probabilities: Sequence[float],
+        *,
+        probability_epsilon: float = 1.0e-5,
+        nonzero_epsilon: float = 1e-6,
+    ) -> 'SparseStatePreparationAliasSampling':
+        """Factory to construct the state preparation gate for a given set of LCU coefficients.
+
+        Args:
+            lcu_probabilities: The LCU coefficients.
+            probability_epsilon: The desired accuracy to represent each probability
+                (which sets mu size and keep/alt integers).
+                See `qualtran.linalg.lcu_util.preprocess_lcu_coefficients_for_reversible_sampling`
+                for more information.
+            nonzero_epsilon: minimum value for a probability entry to be considered non-zero.
+        """
+        alt, keep, mu = preprocess_lcu_coefficients_for_reversible_sampling(
+            lcu_coefficients=lcu_probabilities, epsilon=probability_epsilon
+        )
```

Comment:
Are you sure we should be passing `lcu_probabilities` as a dense array on the original `n` qubit state instead of a dense array on a smaller `log(S)` qubit state; where `S` is the sparsity; to compute the `alt`, `keep` and `mu` indices? Can you verify if the results are same / different and which is the right thing to do?

Reply:
Results are same in both cases, but passing the smaller state is better. I updated the factory method to pass only non-zero values now.